src/vmm/src/device_manager/mod.rs

```diff
@@ -15,3 +47,422 @@ pub mod mmio;
 pub mod persist;
 /// Resource manager for devices.
 pub mod resources;
+
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+/// Error while creating a new [`DeviceManager`]
+pub enum DeviceManagerCreateError {
+    /// Error with EventFd: {0}
+    EventFd(#[from] std::io::Error),
+    #[cfg(target_arch = "x86_64")]
+    /// Legacy device manager error: {0}
+    PortIOError(#[from] LegacyDeviceError),
+    /// Resource allocator error: {0}
+    ResourceAllocator(#[from] vm_allocator::Error),
+}
+
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+/// Error while attaching a VirtIO device
+pub enum AttachMmioDeviceError {
+    /// MMIO transport error: {0}
+    MmioTransport(#[from] MmioError),
+    /// Error inserting device in bus: {0}
+    Bus(#[from] vm_device::BusError),
+}
+
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+/// Error while attaching the VMGenID device
+pub enum AttachVmgenidError {
+    /// Error creating VMGenID device: {0}
+    CreateVmGenID(#[from] VmGenIdError),
+    /// Error while registering VMGenID with KVM: {0}
+    AttachVmGenID(#[from] kvm_ioctls::Error),
+}
+
+#[cfg(target_arch = "aarch64")]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+/// Error while attaching the VMGenID device
+pub enum AttachLegacyMmioDeviceError {
+    /// Cmdline error
+    Cmdline,
+    /// Error creating serial device: {0}
+    CreateSerial(#[from] std::io::Error),
+    /// Error registering device: {0}
+    RegisterMMIODevice(#[from] MmioError),
+    /// Error inserting device in the Bus: {0}
+    Bus(#[from] vm_device::BusError),
+}
```

Comment:
Can this be combined into singe enum?

Reply:
What would the benefit be? I tend to prefer having each function return only the errors that can happen in the function.

Reply:
I see your point. At the same time it would be less error enum juggling if there are less of them. Also all of these errors are critical (e.g. FC exits if any of they appear), so there is an argument to not have any of them. But this is a discussion for another time. Not going to insist on this merge.